iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp

```diff
@@ -627,12 +627,14 @@ static LogicalResult getPoolingOpLaunchConfig(
   // be able to figure out which dimensions of the output correspond to the
   // pooled dimension and which are not. Need to fix that, but for now just use
   // a working heuristic.
-  SmallVector<int64_t, 4> ts(std::min<int64_t>(
-      op.getOutput(0).getType().template cast<ShapedType>().getRank(), 3));
   const int64_t tileSizeX = 32;
   int64_t tileSizeY = maxWorkgroupSize / tileSizeX;
-  ts[ts.size() - 2] = tileSizeY;
-  ts[ts.size() - 1] = tileSizeX;
+  SmallVector<int64_t, 4> ts;
+  if (options.usingLinalgOnTensors) {
+    ts.assign({0, tileSizeY, tileSizeX, 1});
```

Comment:
Yes, Hanhan, you are right about the fact that this is fairly haphazard. Its on my list of things to clean up. 

The overall goal here is that you only want to partition upto 3 loops at the Flow level. The HAL layer can only support 3 levels of processor x,y,z. We can have Flow in theory handle any number of parallelism mode, but it gets tricky then to lower from Flow -> HAL. So for now the assumption is to try to keep 3 levels of parallelism at flow level.

W.R.T to how to decide which loops to parallelize and which not to parallelize, and what tile size to use. We should uniformly try to say tile size of 0 implies it is not parallelized, both in the Flow level when the distrubution happens, and here when setting the configuration.

Let me look through your change, in a bit. Its the thing I want to get to next, so I will have to see whats done currently and then give you feedback here.

Reply:
It's complicated because of three main reasons: 1) existing buffer vs. new tensor path, 2) convolution/pooling ops has more than 3 parallel dimensions but right now we can only support <= 3 Flow tiling, 3) the tile sizes here matches linalg op implicit loop order, which is typically the reverse of distribution order.

In the existing buffer path by default we always tile along the *first* three parallel dimensions, namely N, OH, OW. So that's why you see we only have three numbers here for the existing buffer path: they are for the first 3 implicit loop of conv/pooling ops, namely N, OH, OW. That's not very good because N is typically 1 in inference so we are effectively dropping one dimension of parallelism. That's why we want to tile along the *second* three parallel dimensions, namely OH, OW, OC. The new tensor based approach adopts this scheme for conv, by having the special check in DispatchLinalgOnTensorsPass to force tiling along the *second* three parallel dimensions, namely OH, OW, OC. Now remember that these tiling configurations need to match with implicit loop order in linalg ops. So we need a special leading 0 to mean disabling tiling on the first parallel dimension, namely N. Given you made pooling ops be treated similarly like convolution ops, you need the same thing here. Hope this clarifies. :)

It will improve a bit once we can remove the existing buffer path. But the special checks in DispatchLinalgOnTensorsPass will remain. I would actually think making flow level workgroups really n-D (n being arbitrary number) good to address the special checks. I was doing that in the initial prototype. Will send out a pull request later to show how it looks and whether we should proceed with it.

Reply:
Now I understand more about these, thanks for the detailed explanantion!